File: web_version/pmugg_files/shapes3D/faceType3D.js
ms.faceType3D = function(material, normal) {
	this.material = material;
	this.normal = normal;
	var {u, v} = normal.orthonormal();
	this.u = u;
	this.v = v;
	this.monotonic = false;
	this.color = null;

	// Compute the maximum dimension of the normal.
	this.maxDim = ms.util.maxDim(normal);

	this.id = ms.faceType3D.count++;
};

ms.faceType3D.count = 0;

ms.faceType3D.EPS = 1e-5;

ms.faceType3D.prototype.export = function() {
	return {
		color: this.color,
		material: this.material,
		normal: this.normal.export(),
		// maxDim: this.maxDim,
		// u: this.u.export(),
		// v: this.v.export(),
	}
};

ms.faceType3D.import = function(json) {
	var result = new ms.faceType3D(json.material, ms.vec3.import(json.normal));
	result.color = json.color;
	return result;
};

ms.faceType3D.partialImport = function(json) {
	var normal = ms.vec3.import(json.normal).swapAxes();
	var result = new ms.faceType3D(null, normal);
	result.id = json.signature;
	result.idNum = json.id;
	return result;
};

ms.faceType3D.prototype.getMaterial = function() {
	return this.material;
};

ms.faceType3D.prototype.getNormal = function() {
	return this.normal;
};

ms.faceType3D.prototype.getMaxDim = function() {
	return this.maxDim;
};

// Find the angle within the face's coordinate system.
ms.faceType3D.prototype.angle = function(q) {
	var dx = this.u.dot(q);
	var dy = this.v.dot(q);
	var angle = Math.atan2(dy, dx);
	// Angles near pi wrap to -pi.
	if (angle > Math.PI - ms.faceType3D.EPS) {
		return -angle;
	} else {
		return angle;
	}
};

// Find the area of a face.
ms.faceType3D.prototype.getArea = function(model, face) {	
	var n = face.vertices.length;
	var vProjected = [];
	for (var i = 0; i < n; i++) {
		var vertex = model.vertices[face.vertices[i].vertexIndex];
		vProjected.push([this.u.dot(vertex), this.v.dot(vertex)]);
	}
	return -ms.shapeSet.polygonArea(vProjected);
};

ms.faceType3D.prototype.computeMonotonic = function(graphs) {
	var positive = false;
	var negative = false;
	var self = this;
	graphs.forEach(function(graph) {
		graph.getOuterEndpoints().forEach(function(endpoint) {
			if (endpoint.faceTypes().includes(self) && endpoint.oriented(self)) {
				var trace = ms.graph.traceToExit3D(endpoint, self);
				if (trace.winding > 0) {
					positive = true;
				}
				if (trace.winding < 0) {
					negative = true;
				}
			}
		});
	});
	this.monotonic = positive ^ negative;
};

ms.faceType3D.prototype.normalColor = function(normal) {
	if (this.color) {
		return this.color;
	}
	var normal = this.normal;
	// return [-0.35 * normal.x + 0.5, -0.3 * normal.y + 0.5, 0.2 * normal.z + 0.5, 1];
	return [0.5 * normal.x + 0.5, 0.5 * normal.y + 0.5, 0.5 * normal.z + 0.5, 1];
};

File: cpp_version/pmugg/shapes3D/face_type3d.cpp
#include "pch.h"
#include "face_type3d.h"
// #include "graph.h"
// #include "endpoint.h"
#include "../util/util.h"
#include "../shape/vec2.h"
#define _USE_MATH_DEFINES
#include <cmath>
#include <math.h>

namespace ms {

int FaceType3D::nextId = 0;

FaceType3D::FaceType3D(const std::string& mat, const Vec3& n)
    : material(mat)
    , normal(n)
    , monotonic(false)
    , color(nullptr)
    , id(nextId++) {
    computeOrthonormalBasis();
    maxDim = Util::maxDim(normal);
}

void FaceType3D::computeOrthonormalBasis() {
    // Find the least significant component of the normal
    int minDim = 0;
    float minVal = std::abs(normal[0]);
    for (int i = 1; i < 3; i++) {
        if (std::abs(normal[i]) < minVal) {
            minDim = i;
            minVal = std::abs(normal[i]);
        }
    }

    // Create a vector orthogonal to the normal
    u = Vec3();
    u.setValue(1.0f, minDim);
    u = normal.cross(u).normalize();

    // Complete the orthonormal basis
    v = normal.cross(u).normalize();
}

float FaceType3D::angle(const Vec3& q) const {
    float dx = u.dot(q);
    float dy = v.dot(q);
    float angle = std::atan2(dy, dx);
    
    // Angles near pi wrap to -pi
    if (angle > M_PI - EPS) {
        return -angle;
    }
    return angle;
}

float FaceType3D::getArea(const std::vector<Vec3>& vertices) const {
    std::vector<Vec2> projectedVertices;
    for (const auto& vertex : vertices) {
        projectedVertices.push_back({u.dot(vertex), v.dot(vertex)});
    }
    return -polygonArea(projectedVertices);
}

float FaceType3D::polygonArea(const std::vector<Vec2>& points) const {
    float area = 0;
    for (size_t i = 0; i < points.size(); i++) {
        const auto& p1 = points[i];
        const auto& p2 = points[(i + 1) % points.size()];
        area += p1.getX() * p2.getY() - p2.getX() * p1.getY();
    }
    return area / 2;
}

//void FaceType3D::computeMonotonic(const std::vector<Graph*>& graphs) {
//    bool positive = false;
//    bool negative = false;
//
//    for (const auto* graph : graphs) {
//        for (const auto* endpoint : graph->getOuterEndpoints()) {
//            if (endpoint->getFaceTypes().contains(this) && endpoint->oriented(this)) {
//                auto trace = graph->traceToExit3D(endpoint, this);
//                if (trace.winding > 0) {
//                    positive = true;
//                }
//                if (trace.winding < 0) {
//                    negative = true;
//                }
//            }
//        }
//    }
//
//    monotonic = positive ^ negative;  // XOR operation
//}

Vec3 FaceType3D::normalColor() const {
    if (color) {
        return *color;
    }
    return Vec3(
        0.5f * normal.getX() + 0.5f,
        0.5f * normal.getY() + 0.5f,
        0.5f * normal.getZ() + 0.5f
    );
}

/* Json FaceType3D::export() const {
    Json json;
    json["color"] = color ? color->export() : nullptr;
    json["material"] = material;
    json["normal"] = normal.export();
    return json;
} */

FaceType3D* FaceType3D::import(const Json& json) {
    std::string material = "";
    if (json.contains("material") && json["material"].is_string()) {
        material = json["material"].get<std::string>();
    }
    auto normal = json.contains("normal") ?
        Vec3::import(json["normal"]) : Vec3(0, 0, 1);
    auto* result = new FaceType3D(material, normal);
    
    if (json["color"] != nullptr) {
        result->color = new Vec3(Vec3::import(json["color"]));
    }
    
    return result;
}

FaceType3D* FaceType3D::partialImport(const Json& json) {
    auto* result = new FaceType3D(
        "",  // Empty material for partial import
        Vec3::import(json["normal"]).swapAxes()
    );
    result->id = json["signature"];
    return result;
}

} // namespace ms 